src/native/libs/System.Native/pal_interfaceaddresses.c

```diff
@@ -95,14 +100,78 @@ static inline uint8_t mask2prefix(uint8_t* mask, int length)
 
     return len;
 }
+
+#if !HAVE_IFADDRS && TARGET_ANDROID
+// This structure is exactly the same as struct ifaddrs defined in ifaddrs.h but since the header
+// might not be available (e.g., in bionics used in Android before API 24) we need to mirror it here
+// so that we can dynamically load the getifaddrs function and use it.
+struct ifaddrs
+{
+	struct ifaddrs *ifa_next;
+	char *ifa_name;
+	unsigned int ifa_flags;
+	struct sockaddr *ifa_addr;
+	struct sockaddr *ifa_netmask;
+	union
+	{
+		struct sockaddr *ifu_broadaddr;
+		struct sockaddr *ifu_dstaddr;
+	} ifa_ifu;
+	void *ifa_data;
+};
+#endif
+
+#if !HAVE_GETIFADDRS && TARGET_ANDROID
+// Try to load the getifaddrs and freeifaddrs functions manually.
+// This workaround is necessary on Android prior to API 24 and it can be removed once
+// we drop support for earlier Android versions.
+static int (*getifaddrs)(struct ifaddrs**) = NULL;
+static void (*freeifaddrs)(struct ifaddrs*) = NULL;
+
+static bool ensure_getifaddrs_is_loaded()
+{
+    static bool loading_already_attempted = false;
+    static pthread_mutex_t lock = PTHREAD_MUTEX_INITIALIZER;
+
+    if (!loading_already_attempted)
+    {
+        if (pthread_mutex_lock(&lock) == 0)
+        {
+            if (!loading_already_attempted)
+            {
+                loading_already_attempted = true;
```

Comment:
The race condition is still here.

Reply:
Oh, am I using `pthread_mutex_lock` incorrectly? I assumed using it this way is sufficient (and we're using it like this for example in `pal_console.c`). Or do you mean the case when locking fails and we skip the initialization?

Reply:
I assume you meant `SystemNative_InitializeTerminalAndSignalHandling` in `pal_console.c`. The code in `SystemNative_InitializeTerminalAndSignalHandling` does not read the `initialized` flag outside the lock and it sets the `initialized` as the very last thing (not strictly necessary, but good general practice). 

Your code has race condition because it reads `loading_already_attempted` outside the lock.

Reply:
I'm reading the variable outside and also inside of the lock (to prevent unnecessary locking). I assumed that `pthread_mutex_lock` synchronizes memory, but maybe that's not actually guaranteed then. I'll modify the code to only read it inside and to set the value at the end👍